### ISSUE TITLE
[Discussion] Information about naming conventions for classes affected by pylint "no-self-argument" errors

### ISSUE THREAD

**Issue by lbianchi-lbl** [IDAES/idaes-pse#84]
This discussion/question is part of the ongoing effort described in IDAES/idaes-dev-issues#23 to eliminate pylint false-positive errors in the current codebase.

### Why is pylint complaining?

Similarly to what described in IDAES/idaes-dev#1220, pylint considers it a `no-self-argument` error whenever the first argument in a method definition is not named `self`. This in turn causes a large amount of false-positive errors being reported by pylint for many classes that function perfectly well at runtime.

In addition to the affected classes (and proposed fixes) described in IDAES/idaes-dev#1220, the main category of class where these false-positive errors occur is in `Block`/`BlockData` subclasses.

Taking as an example the definition of `SaponificationStateBlockData` in `idaes.generic_models.properties.examples.saponification_thermo` ([link](https://github.com/IDAES/idaes-dev/blob/7329d04f478ed20cf52c900647a28b306af43805/idaes/generic_models/properties/examples/saponification_thermo.py#L224)), we see that `self`, `blk`, and `b` are all used as first argument for different methods within the same class.

### Possible solutions

In this case, unlike in IDAES/idaes-dev#1220, I'd be more inclined to address these issues at the pylint/syntax level than by making changes to the code. The "blunt scalpel" option of disabling the `no-self-argument` globally in pylint is always on the table, but more fine-grained exclusions would be preferable (if feasible/applicable), since they would allow pylint to be used as a (machine-enforceable) way to make sure that existing practices are respected.

### My questions

To be able to suggest an appropriate strategy to address these pylint issues, I wanted to ask a few general questions to get a better sense of what are the requirements, conventions, and design considerations involved:

1. What are the naming conventions for the first argument of a method defined in one of the affected classes?
1. Is there a difference (based on their first argument's name) in how the methods are called at runtime?
1. Is there a semantic criteria why the first argument is named `self`, `blk`, or `b`?
1. Are `blk` and `b` the only non-`self` names used?
1. Are the "non-`self` methods" (method definitions where the first argument is named something other than `self`) specific to a specific kind of classes (e.g. only for `ProcessBlock`/`ProcessBlockData`, but not other kinds of `Block`/`BlockData` classes?)

Essentially, I'd be interested to know more about existing practices or logic in the codebase in this regard. Based on the `git blame`s I'm paging @eslickj and @andrewlee94 but of course everyone's thoughts on this are very welcome.

**Comment by ksbeattie** [comment]
A final decision on this will happen after pr #58 is merged.

**Comment by lbianchi-lbl** [comment]
This could be done as part of the IDAES v2 codebase refactoring/improvements.

**Comment by andrewlee94** [comment]
@lbianchi-lbl I agree - we should take this opportunity to fix the codebase to resolve this. However, we would need to discuss what is required (i.e. where we should use `self`, and where we shouldn't).

**Comment by eslickj** [comment]
@andrewlee94, I think the answer is that we should use self everywhere pylint complains (first argument to methods in a class where self is an object).
